megabus/src/main/java/com/bazaarvoice/megabus/refproducer/MegabusRefProducerManager.java

```diff
@@ -0,0 +1,91 @@
+import org.apache.curator.framework.CuratorFramework;
+
+import static java.util.Objects.requireNonNull;
+
+public class MegabusRefProducerManager extends PartitionedLeaderService {
+
+    private static final String LEADER_DIR = "/leader/partitioned-megabus-ref-producer";
+    private static final String SERVICE_NAME = "megabus-ref-producer";
+
+    private final Databus _databus;
+    private final int _numRefPartitions;
+    private final String _applicationId;
+
+    @Inject
+    public MegabusRefProducerManager(LeaderServiceTask leaderServiceTask,
+                                     @MegabusZookeeper CuratorFramework curator,
+                                     @MegabusRefTopic Topic refTopic,
+                                     @SelfHostAndPort HostAndPort hostAndPort,
+                                     Clock clock,
+                                     final MegabusRefProducerConfiguration refProducerConfiguration,
+                                     final DatabusFactory databusFactory,
+                                     DatabusEventStore databusEventStore,
+                                     final @SystemIdentity String systemId,
+                                     @MegabusApplicationId String applicationId,
+                                     final RateLimitedLogFactory logFactory,
+                                     @NumRefPartitions int numRefPartitions,
+                                     final MetricRegistry metricRegistry,
+                                     KafkaCluster kafkaCluster,
+                                     ObjectMapper objectMapper) {
+
+        // Since the megabus reads from the databus's internal event store. We forward the refs directly onto a
+        // kafka-based ref topic. We use a partitioned leader service to evenly balance the partitions among the megabus
+        // servers.
+        super(curator, LEADER_DIR, hostAndPort.toString(),
+                "PartitionedLeaderSelector-MegabusRefProducer", numRefPartitions, 1, 1, TimeUnit.MINUTES,
+                partition -> new MegabusRefProducer(refProducerConfiguration, databusEventStore,
+                                logFactory, metricRegistry, kafkaCluster.producer(), objectMapper, refTopic,
+                                ChannelNames.getMegabusRefProducerChannel(applicationId, partition), Integer.toString(partition)), clock);
+
+        _databus = requireNonNull(databusFactory).forOwner(systemId);
+        _numRefPartitions = numRefPartitions;
+        _applicationId = requireNonNull(applicationId);
+
+        for (LeaderService leaderService : getPartitionLeaderServices()) {
+            ServiceFailureListener.listenTo(leaderService, metricRegistry);
+        }
+
+        leaderServiceTask.register(SERVICE_NAME, this);
+    }
+
+    public void createRefSubscriptions() {
+        // TODO: since partitioned databus subscriptions are 1-based, we must add one to the partition condition. At some point in the future,
```

Comment:
are we going to address this in this PR?

Reply:
We are not, as it could break backwards compatibility with some of our databus consumers.